pallets/asset/src/lib.rs

```diff
@@ -2389,4 +2151,131 @@ impl<T: Trait> Module<T> {
+
+        // Ensure its registered by DID or at least expired, thus available.
+        let available = match Self::is_ticker_available_or_registered_to(&ticker, did) {
+            TickerRegistrationStatus::RegisteredByOther => {
+                return Err(Error::<T>::TickerAlreadyRegistered.into())
+            }
+            TickerRegistrationStatus::RegisteredByDid => false,
+            TickerRegistrationStatus::Available => true,
+        };
+
+        if !divisible {
+            ensure!(
+                total_supply % ONE_UNIT.into() == 0.into(),
+                Error::<T>::InvalidTotalSupply
+            );
+        }
+
+        let token_did = <identity::Module<T>>::get_token_did(&ticker)?;
+        // Ensure there's no pre-existing entry for the DID.
+        // This should never happen, but let's be defensive here.
+        identity::Module::<T>::ensure_no_id_record(token_did)?;
+
+        // Charge protocol fees.
+        T::ProtocolFee::charge_fees(&{
+            let mut fees = arrayvec::ArrayVec::<[_; 2]>::new();
+            if available {
+                fees.push(ProtocolOp::AssetRegisterTicker);
+            }
+            // Waive the asset fee iff classic ticker hasn't expired,
+            // and it was already created on classic.
+            if available
+                || ClassicTickers::get(&ticker)
+                    .filter(|r| r.is_created)
+                    .is_none()
+            {
+                fees.push(ProtocolOp::AssetCreateAsset);
+            }
+            fees
+        })?;
+
+        //==========================================================================
+        // At this point all checks have been made; **only** storage changes follow!
+        //==========================================================================
+
+        identity::Module::<T>::commit_token_did(token_did, ticker);
+
+        // Register the ticker or finish its registration.
+        if available {
+            // Ticker not registered by anyone (or registry expired), so register.
+            Self::_register_ticker(&ticker, did, None);
+        } else {
+            // Ticker already registered by the user.
+            <Tickers<T>>::mutate(&ticker, |tr| tr.expiry = None);
+        }
+
+        let token = SecurityToken {
+            name,
+            total_supply,
+            owner_did: did,
+            divisible,
+            asset_type: asset_type.clone(),
+            primary_issuance_agent: Some(did),
+        };
+        <Tokens<T>>::insert(&ticker, token);
+
+        // If asset is not confidential mint assets to the primary asset issuer account here.
+        if !is_confidential {
+            <BalanceOf<T>>::insert(ticker, did, total_supply);
+            Portfolio::<T>::set_default_portfolio_balance(did, &ticker, total_supply);
+        }
+        <AssetOwnershipRelations>::insert(did, ticker, AssetOwnershipRelation::AssetOwned);
+        Self::deposit_event(RawEvent::AssetCreated(
```

Comment:
I think this also needs to be under if confidential condition.

Reply:
Good call.

Reply:
@satyamakgec Why do you think so? There is nothing confidential in the event. Perhaps we can add a Confidential/Non-Confidential enum to the event but I see no reason to not emit an event here.

Reply:
Ah, confidential lib is emitting a different event. I'd suggest that we use this event for both confidential and non-confidential assets with an enum differentiating between both.

Reply:
I did not want to emit that event to create a differentiation b/w confidential and non-confidential asset creation but if we choose the enum to represent the asset type then I think it makes sense and we can emit `AssetCreated` event in the creation of the confidential asset.